### PR TITLE
UCP/WIREUP: Retry getting CM initial config if it doesn't fit CM private data

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Retry getting CM initial config if it doesn't fit CM private data.

## Why ?

Fixes:
```
$ UCX_TLS=ud,rc,sm gdb --args ./src/tools/info/.libs/ucx_info -ep -u tme -n 128 -A 1.1.3.28
...
[1617052881.795899] [jazz28:155982:async]       wireup_cm.c:427  UCX  ERROR CM private data buffer is too small to pack UCP endpoint info, ep 0x7fffece17000/0x7fffece17078 service data 11, address length 64, cm 0xb955f0 max_conn_priv 54
[1617052881.795937] [jazz28:155982:async]      ucp_worker.c:560  UCX  ERROR ep 0x7fffece17000: error 'Provided buffer is too small' on CM lane will not be handled since no error callback is installed
<Failed to flush UCP endpoint>
```
i.e. address of a CM initial configuration doesn't fit CM private data.

## How ?

If the initial config doesn't fit CM private data, retry getting CM initial config with a single transport selected (i.e. just the first one from TL bitmap):
1. set `single_tl = 0`, `log_level = UCS_LOG_LEVEL_DEBUG`
2. get CM initial config
3. try to pack
4. if packing fails, set `single_tl = 1`, `log_level = UCS_LOG_LEVEL_ERROR`
5. get CM initial config using only single transport
6. try to pack

### CODE REVIEW

**Comment by dmitrygx** [comment]
bot:pipe:retest

**Comment by dmitrygx** [comment]
@brminich @evgeny-leksikov could you review pls?

File: src/ucp/wireup/wireup_cm.c [hunk 1]
```diff
@@ -158,9 +158,36 @@ ucp_cm_tl_bitmap_get_dev_idx(ucp_context_h context, ucp_tl_bitmap_t tl_bitmap)
     return dev_index;
 }
 
+static void
+ucp_cm_tl_bitmap_single_tl(ucp_context_h context, ucp_tl_bitmap_t *tl_bitmap)
```
Comment (evgeny-leksikov) [inline src/ucp/wireup/wireup_cm.c:162]: maybe `*first_tl`? BTW, can we select "best" instead of "first"? then it could be single...

**Review by evgeny-leksikov — COMMENTED** [review]
Could you describe the case? I think this happens due to "UCX_TLS" is defined.
Without it, UCP would fallback to TCP and select TCP TL for first round of wireup. Taking into account "all_devices" feature, IB transports would be selected anyway.
Also, since we have practically unlimited private data in TCP we could pack full worker address and avoid MSG WIREUP protocol to establish the UCP connection.

**Comment by dmitrygx** [comment]
> Could you describe the case?

it happens that UD was selected for doing AM and RC for RNDV - so it was not able to pack it into RDMACM private data which is 54, but 64 is required.

> Without it, UCP would fallback to TCP and select TCP TL for first round of wireup. Taking into account "all_devices" feature, IB transports would be selected anyway.

we don't do fallback if private data is too small to pack an address

**Comment by brminich** [comment]
> we don't do fallback if private data is too small to pack an address

Maybe better do that?

**Comment by dmitrygx** [comment]
> > we don't do fallback if private data is too small to pack an address
> 
> Maybe better do that?

done

File: src/ucp/wireup/wireup_cm.c (hunk 1 above)
Comment (dmitrygx) [reply src/ucp/wireup/wireup_cm.c]: done

**Comment by dmitrygx** [comment]
bot:pipe:retest

**Comment by dmitrygx** [comment]
bot:pipe:retest

**Comment by yosefe** [comment]
failure on ppc seems relevant:
```
[2021-05-05T06:56:51.839Z] [ RUN      ] all/test_ucp_sockaddr_cm_private_data.create_multiple_lanes_have_fallback_to_next_cm/6 <all/stream,not_all_devs>
[2021-05-05T06:56:52.097Z] [     INFO ] Testing 65.65.65.12:0
[2021-05-05T06:56:52.098Z] [     INFO ] server listening on 65.65.65.12:48010
[2021-05-05T06:57:59.821Z] [1620197872.035979] [r-vmb-ppc-jenkins:2574 :0]       ucp_test.cc:237  UCX  ERROR request 0x1000b29c010 did not complete on time
[2021-05-05T06:57:59.821Z] [     INFO ] ignoring error Connection reset by remote peer on endpoint 0x3fff9dd90000
[2021-05-05T06:57:59.821Z] [1620197872.090533] [r-vmb-ppc-jenkins:2574 :0]      ucp_worker.c:2442 UCX  WARN  not all pending operations (4) were flushed on worker 0x1000b08f4d0 that is being destroyed
[2021-05-05T06:57:59.821Z] [1620197872.090596] [r-vmb-ppc-jenkins:2574 :0]           mpool.c:44   UCX  WARN  object 0x1000b29bf00 was not returned to mpool ucp_requests
[2021-05-05T06:57:59.821Z] [1620197872.198500] [r-vmb-ppc-jenkins:2574 :0]        rc_iface.c:712  UCX  WARN  some eps were not destroyed
[2021-05-05T06:57:59.821Z] [1620197872.199062] [r-vmb-ppc-jenkins:2574 :0]  rc_mlx5_common.c:581  UCX  WARN  mlx5dv_devx_obj_destroy(SRQ) failed: Invalid argument
[2021-05-05T06:57:59.821Z] [1620197872.205801] [r-vmb-ppc-jenkins:2574 :0]       callbackq.c:452  UCX  WARN  0 fast-path and 1 slow-path callbacks remain in the queue
[2021-05-05T06:57:59.821Z] [1620197872.413757] [r-vmb-ppc-jenkins:2574 :0]           mpool.c:44   UCX  WARN  object 0x1000b5cfe80 was not returned to mpool devx dbrec
[2021-05-05T06:57:59.821Z] [1620197872.413790] [r-vmb-ppc-jenkins:2574 :0]           mpool.c:44   UCX  WARN  object 0x1000b5cff00 was not returned to mpool devx dbrec
[2021-05-05T06:57:59.821Z] [1620197872.415704] [r-vmb-ppc-jenkins:2574 :0]       ib_device.c:677  UCX  WARN  async_events_hash not empty
[2021-05-05T06:57:59.821Z] /scrap/jenkins/workspace/ucx-6/contrib/../test/gtest/common/test.cc:353: Failure
[2021-05-05T06:57:59.821Z] Failed
[2021-05-05T06:57:59.821Z] Got 1 errors and 8 warnings during the test
```

**Comment by dmitrygx** [comment]
bot:pipe:retest

**Comment by dmitrygx** [comment]
@brminich @hoopoepg could you review pls?

File: src/ucp/wireup/select.c [hunk 2]
```diff
@@ -1268,6 +1269,10 @@ ucp_wireup_add_rma_bw_lanes(const ucp_wireup_select_params_t *select_params,
     ucp_tl_bitmap_t tl_bitmap;
     uint8_t i;
 
+    if (ep_init_flags & UCP_EP_INIT_CM_PHASE) {
+        return UCS_OK;
+    }
```
Comment (brminich) [inline src/ucp/wireup/select.c:1274]: are these checks (here and before) unrelated fixes?

File: src/uct/tcp/tcp_sockcm.c [hunk 3]
```diff
@@ -175,6 +175,14 @@ UCS_CLASS_INIT_FUNC(uct_tcp_sockcm_t, uct_component_h component,
                               &uct_tcp_sockcm_iface_ops, worker, component,
                               config);
 
+    if (cm_config->priv_data_len < sizeof(uct_tcp_sockcm_priv_data_hdr_t)) {
+        ucs_error("provided CM private data length (%" PRIu64 ") is to small,"
```
Comment (brminich) [inline src/uct/tcp/tcp_sockcm.c:179]: too small

File: src/ucp/wireup/select.c (hunk 2 above)
Comment (dmitrygx) [reply src/ucp/wireup/select.c:1274]: this is related, we allow only AM lane for CM initial phase

File: src/ucp/wireup/wireup_cm.c [hunk 4]
```diff
@@ -98,19 +98,31 @@ static unsigned ucp_cm_client_try_next_cm_progress(void *arg)
...
+static int ucp_cm_client_get_next_cm_idx(ucp_ep_h ep)
+{
+    ucp_worker_h worker          = ep->worker;
+    ucp_rsc_index_t next_cm_idx  = ucp_ep_ext_control(ep)->cm_idx + 1;
+    ucp_rsc_index_t num_cm_cmpts = ucp_worker_num_cm_cmpts(worker);
+
+    return ((next_cm_idx < num_cm_cmpts) &&
+            (worker->cms[next_cm_idx].cm != NULL)) ?
+                   next_cm_idx :
+                   UCP_NULL_RESOURCE;
```
[the remark this answers is not in the document]
Comment (dmitrygx) [reply src/ucp/wireup/wireup_cm.c:110]: yes, this is a clang suggestion

File: src/ucp/wireup/wireup_cm.c [hunk 5]
```diff
@@ -376,21 +392,33 @@ static unsigned ucp_cm_client_uct_connect_progress(void *arg)
...
-                  " max_conn_priv %lu",
-                  ep, sizeof(ucp_wireup_sockaddr_data_t), ucp_addr_size,
-                  worker->cms[cm_idx].cm,
-                  worker->cms[cm_idx].attr.max_conn_priv);
         status = UCS_ERR_BUFFER_TOO_SMALL;
-        goto free_addr;
+
+        log_level = (ucp_cm_client_get_next_cm_idx(ep) != UCP_NULL_RESOURCE) ?
+                            UCS_LOG_LEVEL_DEBUG :
+                            UCS_LOG_LEVEL_ERROR;
```
[the remark this answers is not in the document]
Comment (dmitrygx) [reply src/ucp/wireup/wireup_cm.c:399]: fixed

File: src/ucp/wireup/wireup_cm.c [hunk 6]
```diff
@@ -1082,6 +1109,8 @@ ucp_ep_cm_server_create_connected(ucp_worker_h worker, unsigned ep_init_flags,
         goto out_free_request;
     }
 
+    ucs_assert(ucp_ep_num_lanes(ep) == 2);
+
```
[the remark this answers is not in the document]
Comment (dmitrygx) [reply src/ucp/wireup/wireup_cm.c:1113]: now only 2 (CM and UCT) lanes could be created

File: src/uct/tcp/tcp_sockcm.c (hunk 3 above)
Comment (dmitrygx) [reply src/uct/tcp/tcp_sockcm.c:179]: fixed

**Comment by dmitrygx** [comment]
@brminich could you review pls? your comments were fixed

File: src/ucp/wireup/select.c [hunk 7]
```diff
@@ -1268,6 +1269,10 @@ ucp_wireup_add_rma_bw_lanes(const ucp_wireup_select_params_t *select_params,
     ucp_tl_bitmap_t tl_bitmap;
     uint8_t i;
 
+    if (ep_init_flags & UCP_EP_INIT_CM_PHASE) {
```
Comment (yosefe) [inline src/ucp/wireup/select.c:1272]: pls rename UCP_EP_INIT_CM_PHASE to what it does and not how it's used

File: src/ucp/wireup/select.c [hunk 8]
```diff
@@ -1383,9 +1388,12 @@ ucp_wireup_add_tag_lane(const ucp_wireup_select_params_t *select_params,
     ucp_ep_h ep                          = select_params->ep;
     ucp_wireup_criteria_t criteria       = {0};
     ucp_wireup_select_info_t select_info = {0};
+    unsigned ep_init_flags               = ucp_wireup_ep_init_flags(
+                                                   select_params, select_ctx);
     ucs_status_t status;
 
     if (!(ucp_ep_get_context_features(ep) & UCP_FEATURE_TAG) ||
+        (ep_init_flags & (UCP_EP_INIT_FLAG_MEM_TYPE | UCP_EP_INIT_CM_PHASE)) ||
```
Comment (yosefe) [inline src/ucp/wireup/select.c:1396]: why added UCP_EP_INIT_FLAG_MEM_TYPE?

File: src/ucp/wireup/wireup_cm.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (yosefe) [reply src/ucp/wireup/wireup_cm.c:110]: what if next_idx is NULL, but it's not the last one?

File: src/ucp/wireup/wireup_cm.c [hunk 9]
```diff
@@ -360,6 +374,8 @@ static unsigned ucp_cm_client_uct_connect_progress(void *arg)
         goto out_check_err;
     }
 
+    ucs_assert(ucp_ep_num_lanes(ep) == 2);
```
Comment (yosefe) [inline src/ucp/wireup/wireup_cm.c:377]: pls document

File: src/uct/tcp/tcp_sockcm.c [hunk 10]
```diff
@@ -175,6 +175,14 @@ UCS_CLASS_INIT_FUNC(uct_tcp_sockcm_t, uct_component_h component,
                               &uct_tcp_sockcm_iface_ops, worker, component,
                               config);
 
+    if (cm_config->priv_data_len < sizeof(uct_tcp_sockcm_priv_data_hdr_t)) {
+        ucs_error("provided CM private data length (%" PRIu64 ") is too small,"
```
Comment (yosefe) [inline src/uct/tcp/tcp_sockcm.c:179]: priv_data_len should configure how much provate data is there **in addition to** sizeof(uct_tcp_sockcm_priv_data_hdr_t)

File: src/ucp/wireup/select.c (hunk 7 above)
Comment (dmitrygx) [reply src/ucp/wireup/select.c:1272]: I think this is suitable name, because it indicates that we do EP configuration selection on CM phase

File: src/ucp/wireup/wireup_cm.c [hunk 11]
```diff
@@ -374,6 +377,8 @@ static unsigned ucp_cm_client_uct_connect_progress(void *arg)
         goto out_check_err;
     }
 
+    /* Only two lanes must be created during CM phase (CM lane and TL lane) of
+     * connection setup between two peers */
```
Comment (yosefe) [inline src/ucp/wireup/wireup_cm.c:381]: can we somehow move this to a common place for client and server?

File: src/ucp/wireup/select.c (hunk 7 above)
Comment (yosefe) [reply src/ucp/wireup/select.c:1272]: per offline discussion, keep as is

**Review by yosefe — APPROVED** [review]

**Review by brminich — APPROVED** [review]

**Comment by dmitrygx** [comment]
@yosefe ok to merge?
